evmcore/validation_test.go

```diff
@@ -0,0 +1,774 @@
+				test.opts(testTransactionsOption()))
+			require.Equal(t, ErrTxTypeNotSupported, err)
+		})
+	}
+}
+
+func TestValidateTx_Nonce_RejectsTxWith(t *testing.T) {
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("older nonce/%v", name), func(t *testing.T) {
+			// setup validation context
+			opt := testTransactionsOption()
+			signer := types.NewPragueSigner(big.NewInt(1))
+
+			// set up to reach nonce check
+			setGasPriceOrFeeCap(t, tx, opt.minTip)
+
+			// set nonce lower than the current account nonce
+			currentNonce := uint64(2)
+			setNonce(t, tx, currentNonce-1)
+
+			// sign txs with sender and set current balance for account
+			_, address, signedTx := signTxForTest(t, tx)
+			testDb := newTestTxPoolStateDb()
+			testDb.nonces[address] = currentNonce
+			opt.currentState = testDb
+
+			// validate transaction
+			err := validateTx(signedTx, signer, opt)
+			require.ErrorIs(t, err, ErrNonceTooLow)
+		})
+	}
+}
+
+func TestValidateTx_Value_RejectsTxWith(t *testing.T) {
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("negative value/%v", name), func(t *testing.T) {
+			if isBlobOrSetCode(tx) {
+				t.Skip("blob and setCode transactions cannot have negative value because they use uint256 Value")
+			}
+			setValueToNegative(t, tx)
+			err := validateTx(types.NewTx(tx), types.NewPragueSigner(big.NewInt(1)),
+				testTransactionsOption())
+			require.ErrorIs(t, err, ErrNegativeValue)
+		})
+	}
+}
+
+func TestValidateTx_GasPriceAndTip_RejectsTxWith(t *testing.T) {
+	extremelyLargeN := new(big.Int).Lsh(big.NewInt(1), 256)
+
+	// GasPrice/GasFeeCap tests
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("gas fee longer than 256 bits/%s", name), func(t *testing.T) {
+			if isBlobOrSetCode(tx) {
+				t.Skip("blob and setCode transactions cannot have gas price larger than uint256")
+			}
+			setGasPriceOrFeeCap(t, tx, extremelyLargeN)
+			err := validateTx(types.NewTx(tx), types.NewPragueSigner(big.NewInt(1)),
+				testTransactionsOption())
+			require.ErrorIs(t, err, ErrFeeCapVeryHigh)
+		})
+	}
+
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("gas price lower than base fee/%v", name), func(t *testing.T) {
+			// setup validation context
+			opt := testTransactionsOption()
+			opt.currentBaseFee = big.NewInt(2)
+
+			// gas fee cap should be higher than current gas price
+			setGasPriceOrFeeCap(t, tx, big.NewInt(1))
+
+			// sign txs with sender
+			signer, _, signedTx := signTxForTest(t, tx)
+
+			// validate transaction
+			err := validateTx(signedTx, signer, opt)
+			require.ErrorIs(t, err, ErrUnderpriced)
+		})
+	}
+
+	// GasTipCap test
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("gas tip longer than 256 bits/%s", name), func(t *testing.T) {
+			// Legacy and access list transactions do not conceptually have a tip.
+			if isLegacyOrAccessList(tx) {
+				t.Skip("legacy and access list transactions do not have gas tip")
```

Comment:
While they do not have gas-tips, their gas-price fields should be treated as both, the gas-tip-cap and gas-fee-cap. So also for those, this test could be relevant.

Reply:
the decision to skip them was the result of semantics discussion with @LuisPH3, which started from https://github.com/0xsoniclabs/sonic/pull/154#discussion_r2024696281.